external/storm-cassandra/src/main/java/org/apache/storm/cassandra/bolt/BatchCassandraWriterBolt.java

```diff
@@ -0,0 +1,192 @@
+     */
+    @Override
+    protected void tick() {
+        prepareAndExecuteStatement();
+    }
+
+    public void prepareAndExecuteStatement() {
+        int size = queue.size();
+        if( size > 0 ) {
+            List<Tuple> inputs = new ArrayList<>(size);
+            queue.drainTo(inputs);
+            try {
+                List<PairStatementTuple> psl = buildStatement(inputs);
+
+                int sinceLastModified = updateAndGetSecondsSinceLastModified();
+                LOG.debug(logPrefix() + String.format("Execute cql batches with %s statements after %s seconds", size, sinceLastModified));
+
+                checkTimeElapsedSinceLastExec(sinceLastModified);
+
+                GroupingBatchBuilder batchBuilder = new GroupingBatchBuilder(cassandraConfConfig.getBatchSizeRows(), psl);
+
+                int batchSize = 0;
+                for (PairBatchStatementTuples batch : batchBuilder) {
+                    LOG.debug(logPrefix() + String.format("Writing data to %s in batches of %s rows.", cassandraConfConfig.getKeyspace(), batch.getInputs().size()));
+                    getAsyncExecutor().execAsync(batch.getStatement(), batch.getInputs());
+                    batchSize++;
+                }
+
+                int pending = getAsyncExecutor().getPendingExec();
+                if (pending > batchSize) {
+                    LOG.warn( logPrefix() + String.format("Currently pending tasks is superior to the number of submit batches(%s) : %s", batchSize, pending));
+                }
+                
+            } catch (Throwable r) {
+                LOG.error(logPrefix() + "Error(s) occurred while preparing batch statements");
+                getAsyncHandler().failure(r, inputs);
+            }
+        }
+    }
+
+    private List<PairStatementTuple> buildStatement(List<Tuple> inputs) {
+        List<PairStatementTuple> stmts = new ArrayList<>(inputs.size());
+
+        for(Tuple t : inputs) {
+            List<Statement> sl = getMapper().map(stormConfig, session, t);
+            for(Statement s : sl)
+                stmts.add(new PairStatementTuple(t, s) );
+        }
+        return stmts;
+    }
+
+    private void checkTimeElapsedSinceLastExec(int sinceLastModified) {
+        if(sinceLastModified > tickFrequencyInSeconds)
+            LOG.warn( logPrefix() + String.format("The time elapsed since last execution exceeded tick tuple frequency - %d > %d seconds", sinceLastModified, tickFrequencyInSeconds));
+    }
+
+    private String logPrefix() {
+        return componentID + " - ";
+    }
+
+    public BatchCassandraWriterBolt withTickFrequency(long time, TimeUnit unit) {
+        this.tickFrequencyInSeconds = (int)unit.toSeconds(time);
+        return this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Map<String, Object> getComponentConfiguration() {
+        Config conf = new Config();
+        conf.put(Config.TOPOLOGY_TICK_TUPLE_FREQ_SECS, tickFrequencyInSeconds);
```

Comment:
we should also have a default value that is less than topology.message.timeout.secs